### PR TITLE
io_request: remove ctor overloads of io_request and s/io_request/const io_request/

### PR DESCRIPTION
**Pull request by tchaikov**
The existing io_request constructor overloads are error-prone. better off using a more explicit way to select the expected constructor. In this change:

* remove ctor overloads. and construct `io_request` instances using the `make_*` methods directly.
* use tagged union for holding different op types. so don't try to shoehorn things into fields.
* access the op members via the op returned by `as<op>()`
* add `sub_req_buffer()` and `sub_req_iovec()` helpers to construct io_request which represents a subrange of current io_request.
* s/io_request/const io_request/

Signed-off-by: Kefu Chai <tchaikov@gmail.com>

### CODE REVIEW

**Comment by xemul** [comment]
> they are but error-prone.

Why?

Nonetheless, all the members of io_request are in fact immutable and deserve being marked with `const`s. This patch will prevent this from happening ever.

The reason why it's not yet constified is because not all members are initialized always, and marking them as such would require adding some initialization to constructors that don't need it. And I'm struggling with my internal perfectionist. On one hand this is negligible, especially compared with the rest of the IO delays, on the other hand it's still extra cycles :roll_eyes:

**Comment by tchaikov** [comment]
@xemul hi Pavel, this change is inspired by https://github.com/scylladb/seastar/pull/1248#discussion_r1000589264 . the reason i claimed these overloads were error-prone is: e759440ce0eb632c7024476e8422dcec0e6f4fb9

> on the other hand it's still extra cycles

why? what extra cycles are you referring? if what you mean is the cycles used for copying the returned instance. i believe C++17 ensures that we can have copy elision when calling these factories.

**Comment by xemul** [comment]
> if we have the luxury of using C++20, probably we could use designated initializers without sacrificing the const specifier?

Well, I did this with c++17 (I believe). Somehow with all fields list-initialized. I will try to dig this old branch out next week.

> why? what extra cycles are you referring?

I was really niggling, never mind :)

```c++
class foo {
    int x;
    int y;
    int z;
    foo(int x_) : x(x_) {}
    foo(int x_, int y_, int z_) : x(x_), y(y_), z(z_) {}
}
```

the 1st foo ctror leaves y and z uninitialized and it costs less ticks.

**Comment by tchaikov** [comment]
i see. guess clang amd gcc (depending on what compiler you were using) are kind of permissive here despite that it is a C++20 feature. see https://godbolt.org/z/exWf57bs1 and https://godbolt.org/z/q3eo7jdcP . if we are not in our pedantic mode, (i don't think we are ! as we are already using `std::ranges`!), we can surely start using [designated initializers](https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers).

but on the other hand, if we want to go with the designated initializers approach, we have to initialize all fields with their in-class initializer and the designated initializers. that'd waste some cycles. but, yes, the first ctor is more efficient in this regard. ahh, now i just understand that your point in your previous "on one hand ..." comment ..

**Comment by avikivity** [comment]
Let's resist the urge to go beyond C++17. And I don't think we use std::ranges, do we?

**Comment by avikivity** [comment]
Yet another idea:

struct io_request_read { ... };
struct io_request_readv { ... };
struct io_request_sendmsg { ... };
// etc

struct io_request {
    enum io_request_type type;
    union {
          io_request_read read;
          io_request_readv readv;
          io_request_sendmsg sendmsg;
          // etc
    } u;
};

e.g. don't try to shoehorn things into fields. But it can be tedious to do the conversion.

**Comment by xemul** [comment]
Idea N+1 (not compile-tested)
```c++
    template<operation Op> struct constructor_selector {}
    io_request(constructor_selector<operation::read>, int fd, void* address, ...) ...
    io_request(constructor_selector<operation::readv>, int fd, std::vector<iovev>& iov, ...) ...
    ...
    static io_request make_read(int fd, void* address, ...) {
        return io_request(constructor_selector<operation::read>{}, fd, address, ...);
    }
    static io_request make_readv(int fd, std::vector<iovec>& iov, ...) {
        return io_request(constructor_selector<operation::readv>{}, fd, iov, ...);
    }
```

**Comment by tchaikov** [comment]
> Let's resist the urge to go beyond C++17. And I don't think we use std::ranges, do we? 

Yes, I think so!

```console
$ git grep 'std::ranges' | cut -d':' -f1 | uniq
include/seastar/coroutine/parallel_for_each.hh
tests/unit/coroutines_test.cc
tests/unit/tl-generator.hh
```

**Comment by tchaikov** [comment]
i tried to implement both suggestions

1. use union. i like it as it is more explicit.
2. use constant type. i also like it. as we don't have to define tag classes for each op. and it allows us to have more `const` .

they led me to #1282 which has FTBFS of
```
/usr/bin/clang++-15 -DFMT_SHARED -DSEASTAR_API_LEVEL=6 -DSEASTAR_DEBUG -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_DEFERRED_ACTION_REQUIRE_NOEXCEPT -DSEASTAR_HAS_MEMBARRIER -DSEASTAR_HAVE_AS$
In file included from /home/kefu/dev/seastar/src/core/reactor_backend.cc:26:
In file included from /home/kefu/dev/seastar/include/seastar/core/reactor.hh:64:
/home/kefu/dev/seastar/include/seastar/core/circular_buffer.hh:489:5: error: static assertion failed due to requirement 'std::is_nothrow_move_assignable<seastar::internal::pending_io_request>::value': erase() as$
    static_assert(std::is_nothrow_move_assignable<T>::value, "erase() assumes move assignment does not throw");
    ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/kefu/dev/seastar/include/seastar/core/internal/io_sink.hh:69:21: note: in instantiation of member function 'seastar::circular_buffer<seastar::internal::pending_io_request>::erase' requested here
        _pending_io.erase(_pending_io.begin(), _pending_io.begin() + drained);
                    ^
/home/kefu/dev/seastar/src/core/reactor_backend.cc:181:36: note: in instantiation of function template specialization 'seastar::internal::io_sink::drain<(lambda at /home/kefu/dev/seastar/src/core/reactor_backend$
    size_t to_submit = _r._io_sink.drain([this] (internal::io_request& req, io_completion* desc) -> bool {
                                   ^
In file included from /home/kefu/dev/seastar/src/core/reactor_backend.cc:21:
In file included from /home/kefu/dev/seastar/src/core/reactor_backend.hh:24:
In file included from /home/kefu/dev/seastar/include/seastar/core/future.hh:24:
In file included from /home/kefu/dev/seastar/include/seastar/core/task.hh:24:
```

so it turns out `circular_buffer` requires the elements in it to be `std::is_nothrow_move_assignable`. and the element type in question is `pending_io_request` which is derived from `io_request`.

**Comment by tchaikov** [comment]
probably we can replace the change in this PR with https://github.com/scylladb/seastar/pull/1282/commits/77894e62d51d24a916ff7e9e35d0ce3df071c5c7 ?

File: include/seastar/core/internal/io_request.hh [hunk 1]
```diff
@@ -266,6 +266,18 @@ public:
         return io_request(operation_tag<operation::cancel>{}, fd, addr);
     }
 
+    io_request(io_request&& other) noexcept
+        : _op(other._op) {
+        std::memcpy(static_cast<void*>(this), static_cast<void*>(&other), sizeof(*this));
```
Comment (avikivity) [inline include/seastar/core/internal/io_request.hh:271]: I'm sure violates some language rule. It also pessimizes the circular_buffer use case.

Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.

Comment (tchaikov) [reply include/seastar/core/internal/io_request.hh:271]: >  It also pessimizes the circular_buffer use case.

 i am not sure what exactly you are referring. is it 

- `io_request(io_request&& other)` operator?  the default-generated one is deleted because otherwise it'd be ill-formed. see https://app.circleci.com/pipelines/github/tchaikov/seastar/170/workflows/c025345c-cf68-4005-a92b-cb5cb5e6a809/jobs/1994 . 
- the `std::memcpy()` call? i think all members variable of the union members are `std::is_trivially_copyable<>`, so we should be able to `memcpy()` them, right?
-  `static_cast<void*>()`? without it, we'd have `-Wclass-memaccess`. like:
   ```
   ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type
   ‘ io_request’ with no trivial copy-assignment;
   use copy-initialization instead [-Wclass-memaccess]
   ```

> Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.

i think what @xemul is after is `const` specifier on the members instead. Pavel, am i right? and i am afraid that the value type of `circular_buffer<Value>` can be a const type, if that's one case of the "everything".

Comment (avikivity) [reply include/seastar/core/internal/io_request.hh:271]: > > It also pessimizes the circular_buffer use case.
> 
> i am not sure what exactly you are referring. is it
> 
> * `io_request(io_request&& other)` operator?  the default-generated one is deleted because otherwise it'd be ill-formed. see https://app.circleci.com/pipelines/github/tchaikov/seastar/170/workflows/c025345c-cf68-4005-a92b-cb5cb5e6a809/jobs/1994 .


User-defined move constructors pessimizes circular_buffer since it now needs to call it.

Without the user-defined constructor, it can just memcpy stuff itself.

> * the `std::memcpy()` call? i think all members variable of the union members are `std::is_trivially_copyable<>`, so we should be able to `memcpy()` them, right?

But memcpy to const members is illegal. The reinterpret_cast is also secretly a const_cast.

> * `static_cast<void*>()`? without it, we'd have warnings from `-Wclass-memaccess`. like:
>   ```
>   ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type
>   ‘ io_request’ with no trivial copy-assignment;
>   use copy-initialization instead [-Wclass-memaccess]
>   ```
> 
> > Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.
> 
> i think what @xemul is after is `const` specifier on the members instead. Pavel, am i right? and i am afraid that the value type of `circular_buffer<Value>` can be a const type, if that's one case of the "everything".

Well, it's a struct. I don't see why we have to qualify members of a struct as const, it's not something typically done in C++ except when there's some special member that must be const.

Comment (tchaikov) [reply include/seastar/core/internal/io_request.hh:271]: > > > It also pessimizes the circular_buffer use case.
> > 
> > 
> > i am not sure what exactly you are referring. is it
> > 
> > * `io_request(io_request&& other)` operator?  the default-generated one is deleted because otherwise it'd be ill-formed. see https://app.circleci.com/pipelines/github/tchaikov/seastar/170/workflows/c025345c-cf68-4005-a92b-cb5cb5e6a809/jobs/1994 .
> 
> User-defined move constructors pessimizes circular_buffer since it now needs to call it.
> 
> Without the user-defined constructor, it can just memcpy stuff itself.

agreed. this is sad. without user-defined constructor. it'd be difficult to add the `const` specifiers.

> 
> > * the `std::memcpy()` call? i think all members variable of the union members are `std::is_trivially_copyable<>`, so we should be able to `memcpy()` them, right?
> 
> But memcpy to const members is illegal. The reinterpret_cast is also secretly a const_cast.

yes. the assignment operator just overwrite the immutable members. this practically defeats the purpose of `const`.

> 
> > * `static_cast<void*>()`? without it, we'd have warnings from `-Wclass-memaccess`. like:
> >   ```
> >   ‘void* memcpy(void*, const void*, size_t)’ writing to an object of type
> >   ‘ io_request’ with no trivial copy-assignment;
> >   use copy-initialization instead [-Wclass-memaccess]
> >   ```
> > 
> > > Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.
> > 
> > 
> > i think what @xemul is after is `const` specifier on the members instead. Pavel, am i right? and i am afraid that the value type of `circular_buffer<Value>` can be a const type, if that's one case of the "everything".
> 
> Well, it's a struct. I don't see why we have to qualify members of a struct as const, it's not something typically done in C++ except when there's some special member that must be const.

ahh, i see your point. it's not specifically a certain (set) of member variables in the op. it's about the whole struct. as the op as a whole is immutable. makes sense to me! but i don't have a quick answer which allows us to make this explicit by adding a `const` somewhere.

maybe we can get the first one reviewed and merged?

i am going to drop the 2nd commit and the 3rd one once @xemul acks this.

Comment (tchaikov) [reply include/seastar/core/internal/io_request.hh]: instead of dropping the 2nd commit and the 3rd commit, i followed @avikivity 's comment:
 
> Is const in some struct worth it? We could make everything const by using `const io_request blah` instead.

and replaced them with a better solution -- to literally making everything `const`.

File: include/seastar/core/circular_buffer.hh [hunk 2]
```diff
@@ -26,6 +26,7 @@
 #include <seastar/util/concepts.hh>
 #include <memory>
 #include <algorithm>
+#include <type_traits>
```
Comment (xemul) [inline include/seastar/core/circular_buffer.hh:29]: This patch doesn't belong to "io_request: <something>" set

File: src/core/reactor_backend.cc [hunk 3]
```diff
@@ -178,7 +178,7 @@ aio_storage_context::submit_work() {
     bool did_work = false;
 
     _submission_queue.resize(0);
-    size_t to_submit = _r._io_sink.drain([this] (internal::io_request& req, io_completion* desc) -> bool {
+    size_t to_submit = _r._io_sink.drain([this] (const internal::io_request& req, io_completion* desc) -> bool {
```
Comment (xemul) [inline src/core/reactor_backend.cc:181]: The `io_sink::drain()` should be patched to manipulate const io_requests

File: include/seastar/core/internal/io_request.hh [hunk 4]
```diff
@@ -38,112 +38,244 @@ public:
     enum class operation { read, readv, write, writev, fdatasync, recv, recvmsg, send, sendmsg, accept, connect, poll_add, poll_remove, cancel };
 private:
     operation _op;
-    int _fd;
```
Comment (xemul) [inline include/seastar/core/internal/io_request.hh:41]: Why moving `_fd` out of "global" location? All operations use it

File: src/core/io_queue.cc [hunk 5]
```diff
@@ -404,26 +404,27 @@ std::vector<io_request::part> io_request::split(size_t max_length) {
 
 std::vector<io_request::part> io_request::split_buffer(size_t max_length) {
     std::vector<part> ret;
-    ret.reserve((_size.len + max_length - 1) / max_length);
+    const auto& op = _read;
```
Comment (xemul) [inline src/core/io_queue.cc:407]: This caught my eye. The `split_buffer` is called for `operation::read` and `operation::write` and this place relies on three assumptions:

0. `read_op _read` and `write_op _write` are unioned. Not that it's going to change ever, but this statement is needed for correctness
1. `using write_op = read_op` exists. This is worse, because compiler won't warn us if some time it changes, e.g. -- zns patchset extends write_op to carry append data on board
2. given 1 is true, compiler lays out `read_op _read` and `write_op _write` matching each other bit-to-bit AND accessing `_read` after initialization took place over `_write` is OK. I don't know standard/compiler well enough to judge how OK is that

To sum up -- item 2 is of the greatest concern on my side

Comment (tchaikov) [reply src/core/io_queue.cc:410]: @xemul agreed. i agree that the 2nd assumption is fragile. so a `static_assert()` should do help us make this explicit and fail the build if the assumption does not hold anymore.

**Comment by tchaikov** [comment]
changelog

- add `static_assert()` to ensure the types of `_read` and `_write`, `_readv` and `_writev` are identical respectively.

File: include/seastar/core/internal/io_request.hh (hunk 4 above)
Comment (tchaikov) [reply include/seastar/core/internal/io_request.hh:41]: yes, all operations use `_fd`. but for the sake of consistency, i am inclined to put all operands of an operation into its own structure. let's think this way, before this change, we try to extract the common parts of operands into a shared union. after this change, all the shared bits are restructured into operation structure itself. `fd` is not an exception in this regard even it is shared by *all* operations.

File: src/core/reactor_backend.cc (hunk 3 above)
Comment (tchaikov) [reply src/core/reactor_backend.cc:181]: there is a dedicated commit to implement the changes in `io_sink`. see 7eb4ec70159e2e60f8d9be0ed40636af85fa3afc

File: include/seastar/core/circular_buffer.hh (hunk 2 above)
Comment (tchaikov) [reply include/seastar/core/circular_buffer.hh:29]: are you referring 

- 96ff6ff041c43009874d62317a45f9a6916f7478 (core/circular_buffer: allow `const` specifier in value_type) ? if that's the case. and
- 64e404b0c6433e90b774fb54b41d971d21108fe8 (core/circular_buffer: refactor loop in circular_buffer::erase())

? 

they are included here just because of https://github.com/scylladb/seastar/pull/1250#issuecomment-1295139142 . so i can assure that the first commit would not prevent us from having `const` semantics of `io_request`.

i'd be happy to move the `core/circular_buffer` changes into a follow up change once we are fine with them. what do you think?

File: src/core/reactor_backend.cc (hunk 3 above)
Comment (xemul) [reply src/core/reactor_backend.cc:181]: Indeed

File: include/seastar/core/circular_buffer.hh (hunk 2 above)
Comment (xemul) [reply include/seastar/core/circular_buffer.hh:29]: Ah, I see. OK, making circular_buffer work with consts is fine to have here. Refactoring -- I'd prefer seeing it as separate PR (unless I'm missing something again and it's also needed to make const-ification compile)

Comment (tchaikov) [reply include/seastar/core/circular_buffer.hh:29]: no, the loop change is not a necessity for adding `const`. but that change has a tiny overlap with  `core/circular_buffer: allow `const` specifier in value_type` change. so it would be very likely it will conflict with this PR. anyway, i agree with you that, strictly speaking, it should belong to another PR. just extracted it into #1283.

File: include/seastar/core/circular_buffer.hh [hunk 6]
```diff
@@ -55,10 +56,10 @@ namespace seastar {
 ///     * pop_back() will invalidate end().
 ///
 /// reserve() may also invalidate all iterators and references.
-template <typename T, typename Alloc = std::allocator<T>>
+template <typename T, typename Alloc = std::allocator<std::remove_const_t<T>>>
```
Comment (avikivity) [inline include/seastar/core/circular_buffer.hh:59]: Ordinary containers don't allow const T:

https://godbolt.org/z/KMbhEYKKW

So I don't think we should either. Nor do I see any gain from it. The code that pops an io_request could pop it into a const variable.

**Comment by tchaikov** [comment]
change log

- dropped the commits to the commits to allow passing `const T` as the template parameter of `circular_buffer`.

@xemul to add more `const`: i implemented following approaches

- to add the tagged ctor and the change: we have to implement the move assignment operator, which overwrites the `const` members. this violates the `const` specifier, and defeats the purpose of adding `const`. and more importantly, what we need is the `const` semantics of the whole structure not member variables of it.
- to enforce `const T` in `circular_buffer`. as Avi pointed out, this is not allowed by ordinary containers. i also feel weird when looking at it. but yeah, if you look at something long enough, there is better chance that it looks strange. but this one is indeed very controversial.

so i gave up on adding more `const` to `io_request`, probably can leave it as an exercise for another raining night.
